### DISCUSSION TITLE
Is there a way  to debug into Node's javascript lib files?

### DISCUSSION THREAD

**Discussion by nextfool** [nodejs/help#4894] [Q&A]
Hi community,  
I am new to Nodejs. I am trying to learn how Node works internally.
Follow the steps in Building.md, I am able to debug into node's (as well as v8, libuv) c++ source code. 
Now I am thinking of stepping into those .js files in /lib directory to have further understanding of how js code working with c++ code.

Given the fact that many of the js files are "pre-compiled" into c++ arrays(or assembly code? please help me to understand here) by jstc.py,  now I am just wondering is it possible to debug the (un-compiled) node js lib files?

Thank you very much.

**Comment by Trott** [comment]
> Given the fact that many of the js files are "pre-compiled" into c++ arrays(or assembly code? please help me to understand here) by jstc.py, now I am just wondering is it possible to debug the (un-compiled) node js lib files?

Yes. Debugging is unaffected by that fact. Even in the CLI debugger, you can step into the internal Node.js JS code.

```text
$  node inspect test.js
< Debugger listening on ws://127.0.0.1:9229/6da937cc-65c6-4ca2-b418-bf54e5452c25
< For help, see: https://nodejs.org/en/docs/inspector
< 
connecting to 127.0.0.1:9229 ... ok
< Debugger attached.
< 
Break on start in test.js:1
> 1 const fs = require("fs");
  2 
  3 fs.readFileSync('/etc/motd');
debug> n
break in test.js:3
  1 const fs = require("fs");
  2 
> 3 fs.readFileSync('/etc/motd');
  4 
  5 
debug> s
break in node:fs:451
 449  */
 450 function readFileSync(path, options) {
>451   options = getOptions(options, { flag: 'r' });
 452   const isUserFd = isFd(path); // File descriptor ownership
 453   const fd = isUserFd ? path : fs.openSync(path, options.flag, 0o666);
debug> 

```
